File: examples/src/python/02_committed.py
#!/usr/bin/env python3
"""
02_committed.py  —  Q2: Committed expenditure (salaries + pensions + interest), Actuals basis.

Run from repo root OR from examples/:
    python3 examples/src/python/02_committed.py

WHAT IS COMMITTED EXPENDITURE?
  Governments have two broad kinds of spending:
    - Committed:  legally/contractually obligated (must pay regardless of budget pressure)
    - Discretionary: can be adjusted year to year

  The three committed components here are:
    Salaries  — direct pay to government employees (specific object head codes)
    Pensions  — post-retirement payments (major head 2071)
    Interest  — payments on state borrowings (major head 2049)

  All three are computed on an 'Actuals' basis (what was actually spent).

PRIORITY RULE:
  pension major head (2071) > interest major head (2049) > salary object codes

  This priority prevents double-counting. Example: a pension payment row has
  major_head_code='2071' AND its object_head_code might also be in the salary
  object codes. Without priority, it would be counted twice. With priority,
  it counts as 'pension' only.

SALARY OBJECT CODES:
  The scope.json lists the exact object head codes that constitute direct
  government salaries (basic pay, allowances, etc.). Grants-in-aid salaries
  (object codes 101, 118) are deliberately excluded — those are payments to
  third-party bodies, not the government's own payroll.

OUTPUT:
  Actuals table (one row per fiscal year) + total_expenditure BE table.
  Both sections are written to out/results_python.json.
"""

import os
import sys
from collections import defaultdict

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import _lib


def main():
    scope = _lib.load_scope()
    rows  = _lib.load_tidy()

    # Scope constants (from scope.json — do not hardcode)
    PENSION_MH  = scope["pension_major_head"]            # "2071"
    INTEREST_MH = scope["interest_major_head"]           # "2049"
    SALARY_OHC  = set(scope["salary_object_codes"])      # {"001", "002", ...}
    L2C         = scope["unit_lakh_to_crore"]            # 100

    # Actuals-basis accumulators
    sal   = defaultdict(float)   # fy -> salary lakh (Actuals)
    pen   = defaultdict(float)   # fy -> pension lakh (Actuals)
    intr  = defaultdict(float)   # fy -> interest lakh (Actuals)
    total = defaultdict(float)   # fy -> total expenditure lakh (Actuals)

    # BE-basis accumulators for the total_expenditure section
    sal_be    = defaultdict(float)
    pen_be    = defaultdict(float)
    int_be    = defaultdict(float)
    total_be  = defaultdict(float)

    for r in rows:
        fy  = r["fiscal_year"]
        mh  = r["major_head_code"]
        oh  = r["object_head_code"]
        amt = r["amount_lakh"]
        m   = r["measure"]

        if m == "Actuals":
            # Every Actuals row contributes to the total
            total[fy] += amt
            # Priority: pension > interest > salary (to prevent double-counting)
            if mh == PENSION_MH:
                pen[fy] += amt
            elif mh == INTEREST_MH:
                intr[fy] += amt
            elif oh in SALARY_OHC:
                sal[fy] += amt
            # All other rows = discretionary; counted in total but not committed

        elif m == "BE":
            total_be[fy] += amt
            # Same priority logic for the BE-based committed figure
            if mh == PENSION_MH:
                pen_be[fy] += amt
            elif mh == INTEREST_MH:
                int_be[fy] += amt
            elif oh in SALARY_OHC:
                sal_be[fy] += amt

    fy_with_act = sorted(total.keys())
    fy_with_be  = sorted(total_be.keys())

    # --- Print the Actuals table ---
    print("Q2: Karnataka Committed Expenditure (Actuals basis) — INR crore")
    print("=" * 95)
    print(
        f"{'Fiscal Year':>12} {'Salaries':>12} {'Pensions':>12} {'Interest':>12}"
        f" {'Committed':>12} {'Total Exp':>12} {'%Share':>7}"
    )
    print("-" * 95)

    q2_rows = []
    for fy in fy_with_act:
        s, p, i = sal[fy], pen[fy], intr[fy]
        committed = s + p + i
        tot       = total[fy]
        share     = round(100.0 * committed / tot, 2) if tot else 0.0

        print(
            f"{fy:>12}"
            f" {s/L2C:>12,.2f}"
            f" {p/L2C:>12,.2f}"
            f" {i/L2C:>12,.2f}"
            f" {committed/L2C:>12,.2f}"
            f" {tot/L2C:>12,.2f}"
            f" {share:>6.2f}%"
        )

        q2_rows.append({
            "fiscal_year":       fy,
            "salaries":          round(s / L2C,         2),
            "pensions":          round(p / L2C,         2),
            "interest":          round(i / L2C,         2),
            "committed":         round(committed / L2C, 2),
            "total_exp_actuals": round(tot / L2C,       2),
            "share_pct":         share,
        })

    print("-" * 95)
    print("  Salaries = direct pay object codes (GIA-salaries excl.).")
    print("  Priority: pensions (MH 2071) > interest (MH 2049) > salaries (by object code).")

    # --- total_expenditure section: BE-based, for all BE years ---
    te_rows = []
    for fy in fy_with_be:
        tb = total_be[fy]
        if tb == 0.0:
            continue
        cb = sal_be[fy] + pen_be[fy] + int_be[fy]
        te_rows.append({
            "fiscal_year":  fy,
            "total_be":     round(tb / L2C, 2),
            "committed_be": round(cb / L2C, 2),
        })

    _lib.record_result("q2_committed",     q2_rows)
    _lib.record_result("total_expenditure", te_rows)
    print("\nRecorded to out/results_python.json  (keys: q2_committed, total_expenditure)")


if __name__ == "__main__":
    main()

File: examples/src/r/02_committed.R
#!/usr/bin/env Rscript
# 02_committed.R — Q2: Committed expenditure — salaries + pensions + interest
#
# "Committed" expenditure is money the government is legally or contractually
# obliged to pay: employee salaries, pensioner payments, and interest on debt.
# It is the least flexible part of the budget.
#
# Categories and their identifiers (from scope.json):
#   Pensions  — major head 2071 (highest priority)
#   Interest  — major head 2049
#   Salaries  — object head codes 001–005, 008–009, 011, 014, 020, 033, 035
#               (only if the row is NOT already counted as pension or interest)
#
# Run from repo root:
#   Rscript examples/src/r/02_committed.R
# ─────────────────────────────────────────────────────────────────────────────

source(file.path(dirname(sub("--file=", "",
  grep("--file=", commandArgs(trailingOnly = FALSE), value = TRUE)[1])),
  "_lib.R"))

# ── Load data ─────────────────────────────────────────────────────────────────
cat("Loading tidy CSV ...\n")
df <- load_tidy()

# ── Compute committed expenditure for each year ───────────────────────────────
# Using Actuals only (we know what was actually paid).
act_yrs <- sort(unique(df$fiscal_year[df$measure == "Actuals"]))

results <- lapply(act_yrs, function(fy) {
  s <- df[df$measure == "Actuals" & df$fiscal_year == fy, ]

  # Priority masks — a row is counted in at most one category.
  is_pension  <- s$major_head_code == PENSION_MH    # 2071
  is_interest <- s$major_head_code == INTEREST_MH   # 2049  (never also pension)
  is_salary   <- s$object_head_code %in% SALARY_OH & !is_pension & !is_interest

  pen  <- sum(s$amount_lakh[is_pension],  na.rm = TRUE)
  intr <- sum(s$amount_lakh[is_interest], na.rm = TRUE)
  sal  <- sum(s$amount_lakh[is_salary],   na.rm = TRUE)

  total_act <- sum(s$amount_lakh, na.rm = TRUE)
  committed <- sal + pen + intr
  share     <- if (total_act > 0) round(100 * committed / total_act, 2) else 0

  list(fy = fy, sal = crore(sal), pen = crore(pen), intr = crore(intr),
       committed = crore(committed), total = crore(total_act), share = share)
})

# ── Print results ─────────────────────────────────────────────────────────────
cat("\nQ2: Committed expenditure — Actuals (INR crore)\n")
cat(sprintf("  %-8s  %10s  %10s  %10s  %10s  %10s  %6s\n",
            "FY", "Salaries", "Pensions", "Interest", "Committed", "Total", "Share%"))
cat(strrep("-", 76), "\n")
for (r in results) {
  cat(sprintf("  %-8s  %10.2f  %10.2f  %10.2f  %10.2f  %10.2f  %5.2f%%\n",
              r$fy, r$sal, r$pen, r$intr, r$committed, r$total, r$share))
}
cat("\nNote: Share% = committed / total Actuals expenditure × 100\n")
cat("Priority rule: pension MH (2071) > interest MH (2049) > salary OH codes\n")

# ── Write complete results_r.json ─────────────────────────────────────────────
cat("\nComputing all sections and writing results_r.json ...\n")
all_results <- compute_all(df)
write_results_json(all_results)
cat("Done.\n")
